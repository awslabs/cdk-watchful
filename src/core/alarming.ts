import { Alarm, AlarmProps, ComparisonOperator, HorizontalAnnotation, IAlarm, IMetric, TreatMissingData } from '@aws-cdk/aws-cloudwatch';
import { Construct } from '@aws-cdk/core';

/**
 * Single alarm customization. If any property is undefined, a default value will be used by the factory.
 */
interface AlarmCustomization {
  /**
   * number of datapoints to alarm on
   * @default factory default
   */
  readonly datapointsToAlarm?: number;
  /**
   * number of periods to evaluate
   * @default factory default
   */
  readonly evaluationPeriods?: number;
  /**
   * whether to enable alarm actions
   * @default factory default
   */
  readonly actionsEnabled?: boolean;
}

/**
 * Single alarm definition with customization.
 */
export interface AlarmDefinition extends AlarmCustomization {
  /**
   * alarm threshold
   */
  readonly threshold: number;
}

/**
 * Properties related to an existing alarm.
 */
export interface CreatedAlarm {
  /**
   * created alarm
   */
  readonly alarm: IAlarm;
  /**
   * annotation of the created alarm
   */
  readonly annotation: HorizontalAnnotation;
  /**
   * the original alarm definition
   */
  readonly definition: AlarmDefinition;
}

/**
 * Factory that creates alarms on metrics.
 */
export interface AlarmFactory {
  /**
   * Creates alarms according to the alarm definitions.
   * @param id alarm identifier (to distinguish alarm from other alarms on the same resource)
   * @param metric metric to alarm on
   * @param op comparison operator
   * @param definitions map of (disambiguator, alarm definition), might be empty or undefined
   */
  createAlarms(id: string, metric: IMetric, op: ComparisonOperator, definitions?: Record<string, AlarmDefinition>): CreatedAlarm[];
}

/**
 * Props to create DefaultAlarmFactory.
 */
export interface DefaultAlarmFactoryProps {
  /**
   * global prefix to be used to all alarms created by this factory
   */
  readonly prefix: string;
  /**
   * default evaluation periods used when there is no custom one
   */
  readonly evaluationPeriodsDefault: number;
  /**
   * default number of datapoints used when there is no custom one
   */
  readonly datapointsToAlarmDefault: number;
  /**
   * default action enabled flag used when there is no custom one
   */
  readonly actionsEnabledDefault: boolean;
}

/**
 * Default implementation of the alarm factory.
 *
 */
export class DefaultAlarmFactory extends Construct implements AlarmFactory {
  protected readonly prefix: string;
  protected readonly evaluationPeriodsDefault: number;
  protected readonly datapointsToAlarmDefault: number;
  protected readonly actionsEnabledDefault: boolean;

  constructor(scope: Construct, id: string, props: DefaultAlarmFactoryProps) {
    super(scope, id);
    this.prefix = props.prefix;
    this.evaluationPeriodsDefault = props.evaluationPeriodsDefault;
    this.datapointsToAlarmDefault = props.datapointsToAlarmDefault;
    this.actionsEnabledDefault = props.actionsEnabledDefault;
  }

  createAlarms(id: string, metric: IMetric, op: ComparisonOperator, definitions?: Record<string, AlarmDefinition>) {
    const createdAlarms: CreatedAlarm[] = [];
    for (let [disambiguator, definition] of Object.entries(definitions ?? {})) {
      createdAlarms.push(this.createAlarm(id, metric, op, disambiguator, definition));
    }
    return createdAlarms;
  }

  protected createAlarm(id: string, metric: IMetric, op: ComparisonOperator, disambiguator: string, props: AlarmDefinition) {
    const alarmId = this.alarmId(id, metric, op, disambiguator, props);
    const alarmName = this.alarmName(id, metric, op, disambiguator, props);
    const alarmDescription = this.alarmDescription(id, metric, op, disambiguator, props);

    const definition: AlarmProps = {
      metric,
      threshold: props.threshold,
      comparisonOperator: op,
      treatMissingData: TreatMissingData.MISSING,
      alarmName,
      alarmDescription,
      evaluationPeriods: props.evaluationPeriods ?? this.evaluationPeriodsDefault,
      datapointsToAlarm: props.datapointsToAlarm ?? this.datapointsToAlarmDefault,
      actionsEnabled: props.actionsEnabled ?? this.actionsEnabledDefault,
    };

    const alarm = new Alarm(this, alarmId, definition);
    const annotation = alarm.toAnnotation();

    return { alarm, definition, annotation };
  }

  protected alarmId(id: string, _metric: IMetric, _op: ComparisonOperator, disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return `${this.prefix}_${id}_${disambiguator}`;
  }

  protected alarmName(_id: string, _metric: IMetric, _op: ComparisonOperator, _disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return undefined;
  }

  protected alarmDescription(_id: string, _metric: IMetric, _op: ComparisonOperator, _disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return undefined;
  }
}
