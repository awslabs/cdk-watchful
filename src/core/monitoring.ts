import { IWidget } from '@aws-cdk/aws-cloudwatch/lib/widget';

/**
 * Represents a monitoring of a single resource or group of related resources.
 * Provides metrics and alarms, and can be subclassed to customize look and feel.
 */
export interface WatchfulMonitoring {
  /**
   * Returns the monitoring widgets.
   * @return widgets
   */
  widgets(): IWidget[];
}
