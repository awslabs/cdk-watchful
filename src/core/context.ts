/**
 * Provides access to global settings and parameters.
 * TODO: currently just a draft
 */
import { Duration } from '@aws-cdk/core';

export class WatchfulContext {
  readonly awsAccountRegion: string;
  readonly defaultMetricPeriod: Duration;

  constructor() {
    // TODO
    this.awsAccountRegion = 'us-east-1';
    this.defaultMetricPeriod = Duration.minutes(5);
  }
}
