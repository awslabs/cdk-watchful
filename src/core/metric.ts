import { DimensionHash, IMetric, MathExpression, Metric } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

/**
 * Similar as CloudWatch Statistic, but also with missing percentiles.
 */
export enum WatchfulMetricStatistic {
  SAMPLE_COUNT = 'SampleCount',
  AVERAGE = 'Average',
  SUM = 'Sum',
  MINIMUM = 'Minimum',
  MAXIMUM = 'Maximum',
  P50 = 'p50',
  P90 = 'p90',
  P99 = 'p99',
}

/**
 * Abstract base class for all Watchful metrics.
 */
export abstract class WatchfulMetric {
  /**
   * Converts the object into the standard CloudWatch metric object.
   */
  abstract unwrap(): IMetric;
}

/**
 * Abstract base class for all metrics that support alarming on them.
 */
export abstract class WatchfulMetricWithAlarmSupport extends WatchfulMetric {
  // empty
}

/**
 * Props to create SimpleWatchfulMetric.
 */
export interface SimpleWatchfulMetricProps {
  readonly namespace: string;
  readonly metricName: string;
  readonly statistic: WatchfulMetricStatistic;
  readonly dimensions?: DimensionHash;
  readonly period: Duration;
  readonly label?: string;
  readonly color?: string;
}

/**
 * Simple metric, corresponding with the emitted datapoints.
 */
export class SimpleWatchfulMetric extends WatchfulMetricWithAlarmSupport {
  protected readonly props: SimpleWatchfulMetricProps;

  constructor(props: SimpleWatchfulMetricProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    return new Metric({
      namespace: this.props.namespace,
      metricName: this.props.metricName,
      dimensions: this.props.dimensions,
      period: this.props.period,
      statistic: this.props.statistic,
      label: this.props.label,
      color: this.props.color,
    });
  }
}

/**
 * Props to create WatchfulMetricMath.
 */
export interface WatchfulMetricMathProps {
  readonly expression: string;
  readonly usingMetrics: Record<string, WatchfulMetric>;
  readonly period: Duration;
  readonly label?: string;
  readonly color?: string;
}

/**
 * Math expression, that transforms different metrics using an expression.
 * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html#metric-math-syntax
 */
export class WatchfulMetricMath extends WatchfulMetricWithAlarmSupport {
  protected props: WatchfulMetricMathProps;

  constructor(props: WatchfulMetricMathProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    const usingMetrics: Record<string, IMetric> = {};

    for (const [key, value] of Object.entries(this.props.usingMetrics)) {
      usingMetrics[key] = value.unwrap();
    }

    return new MathExpression({
      expression: this.props.expression,
      label: this.props.label,
      color: this.props.color,
      period: this.props.period,
      usingMetrics,
    });
  }
}

/**
 * Props to create WatchfulMetricSearch.
 */
export interface WatchfulMetricSearchProps {
  readonly namespace: string;
  readonly dimensions: Record<string, string>;
  readonly searchQuery?: string;
  readonly statistic: WatchfulMetricStatistic;
  readonly period: Duration;
  readonly label: string;
}

/**
 * Search metric results. Does not support creating any kind of alarms.
 * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/search-expression-syntax.html
 */
export class WatchfulMetricSearch extends WatchfulMetric {
  protected props: WatchfulMetricSearchProps;

  constructor(props: WatchfulMetricSearchProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    const period = this.props.period;
    const dimensionKeys = Object.keys(this.props.dimensions).join(',');
    const schema = `{${this.props.namespace},${dimensionKeys}}`;

    function mapper(keyAndValue: string[]) {
      const [key, value] = keyAndValue;
      if (value === '') {
        return key;
      } else {
        return `${key}="${value}"`;
      }
    }

    const dimensionKeysAndValues = Object
      .entries(this.props.dimensions)
      .map(mapper)
      .join(' ');

    const expression = `SEARCH('${schema} ${dimensionKeysAndValues} ${this.props.searchQuery ?? ''}', '${this.props.statistic}', ${period.toSeconds()})`;

    return new MathExpression({
      expression,
      usingMetrics: {},
      label: ' ',
      period,
    });
  }
}
