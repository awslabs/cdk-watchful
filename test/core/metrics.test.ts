import { Duration } from '@aws-cdk/core';
import { SimpleWatchfulMetric, WatchfulMetricMath, WatchfulMetricSearch, WatchfulMetricStatistic } from '../../src/core/metric';

test('simple metric unwrapped', () => {
  // GIVEN

  const metrics = new SimpleWatchfulMetric({
    label: 'DummyLabel',
    statistic: WatchfulMetricStatistic.P99,
    period: Duration.minutes(3),
    dimensions: { A: 'a', B: 'b' },
    metricName: 'DummyMetricName',
    namespace: 'DummyNamespace',
    color: 'DummyColor',
  });

  // WHEN

  const unwrapped = metrics.unwrap();

  // THEN

  expect(unwrapped).toMatchSnapshot();
});

test('metric math unwrapped', () => {
  // GIVEN

  const metric1 = new SimpleWatchfulMetric({
    namespace: 'DummyNamespace',
    metricName: 'DummyMetric1',
    statistic: WatchfulMetricStatistic.AVERAGE,
    period: Duration.hours(1),
  });

  const metric2 = new SimpleWatchfulMetric({
    namespace: 'DummyNamespace',
    metricName: 'DummyMetric2',
    statistic: WatchfulMetricStatistic.AVERAGE,
    period: Duration.hours(1),
  });

  const metrics = new WatchfulMetricMath({
    label: 'DummyLabel',
    period: Duration.minutes(3),
    color: 'DummyColor',
    expression: 'metric1 / metric2',
    usingMetrics: { metric1, metric2 },
  });

  // WHEN

  const unwrapped = metrics.unwrap();

  // THEN

  expect(unwrapped).toMatchSnapshot();
});

test('search metric unwrapped', () => {
  // GIVEN

  const metrics = new WatchfulMetricSearch({
    label: 'DummyLabel',
    statistic: WatchfulMetricStatistic.P99,
    period: Duration.minutes(3),
    dimensions: { A: 'a', B: 'b' },
    namespace: 'DummyNamespace',
    searchQuery: 'DummyQuery',
  });

  // WHEN

  const unwrapped = metrics.unwrap();

  // THEN

  expect(unwrapped).toMatchSnapshot();
});
