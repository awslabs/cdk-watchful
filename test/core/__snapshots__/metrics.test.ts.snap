// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metric math unwrapped 1`] = `
MathExpression {
  "color": "DummyColor",
  "expression": "metric1 / metric2",
  "label": "DummyLabel",
  "period": Duration {
    "amount": 3,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "usingMetrics": Object {
    "metric1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "DummyMetric1",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 3,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Average",
      "unit": undefined,
    },
    "metric2": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "DummyMetric2",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 3,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Average",
      "unit": undefined,
    },
  },
}
`;

exports[`search metric unwrapped 1`] = `
MathExpression {
  "color": undefined,
  "expression": "SEARCH('{DummyNamespace,A,B} A=\\"a\\" B=\\"b\\" DummyQuery', 'p99', 180)",
  "label": " ",
  "period": Duration {
    "amount": 3,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "usingMetrics": Object {},
}
`;

exports[`simple metric unwrapped 1`] = `
Metric {
  "account": undefined,
  "color": "DummyColor",
  "dimensions": Object {
    "A": "a",
    "B": "b",
  },
  "label": "DummyLabel",
  "metricName": "DummyMetricName",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 3,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p99",
  "unit": undefined,
}
`;
